File: forge/lib/tool_contract.py
#!/usr/bin/env python3

import argparse
import json
import sys


class ToolInputError(Exception):
    def __init__(self, code, message):
        super().__init__(message)
        self.code = code


def parse_tool_arguments():
    parser = argparse.ArgumentParser(add_help=True)
    parser.add_argument("--input")
    parser.add_argument("--output")
    return parser.parse_args()


def read_tool_input(args):
    raw = ""
    if args.input:
        with open(args.input, "r", encoding="utf-8") as handle:
            raw = handle.read()
    else:
        raw = sys.stdin.read()
    if not raw.strip():
        return {}
    try:
        return json.loads(raw)
    except json.JSONDecodeError as error:
        raise ToolInputError("invalid_json", str(error)) from error


def ok_result(artifacts=None, warnings=None, data=None):
    return {
        "status": "ok",
        "artifacts": artifacts or [],
        "warnings": warnings or [],
        "errors": [],
        "data": data,
    }


def error_result(error):
    code = getattr(error, "code", "tool_error")
    return {
        "status": "error",
        "artifacts": [],
        "warnings": [],
        "errors": [{"code": code, "message": str(error)}],
    }


def write_tool_result(result, args):
    text = json.dumps(result, indent=2, ensure_ascii=False) + "\n"
    if args.output:
        with open(args.output, "w", encoding="utf-8") as handle:
            handle.write(text)
    else:
        sys.stdout.write(text)


def run_tool(handler):
    args = parse_tool_arguments()
    try:
        write_tool_result(handler(read_tool_input(args)), args)
    except BaseException as error:
        write_tool_result(error_result(error), args)
        raise SystemExit(1) from None


def required_string(payload, key):
    value = payload.get(key)
    if not isinstance(value, str) or not value.strip():
        raise ToolInputError("missing_required_field", f"{key} is required")
    return value

File: forge/lib/tool_contract.mjs
import { readFileSync, writeFileSync } from "node:fs";

export class ToolInputError extends Error {
	constructor(code, message) {
		super(message);
		this.name = "ToolInputError";
		this.code = code;
	}
}

export function parseToolArguments(argv = process.argv.slice(2)) {
	const args = { input: null, output: null };
	for (let index = 0; index < argv.length; index += 1) {
		const argument = argv[index];
		if (argument === "--input") args.input = argv[++index] ?? "";
		else if (argument === "--output") args.output = argv[++index] ?? "";
		else throw new ToolInputError("unknown_option", `Unknown option: ${argument}`);
	}
	if (args.input === "") throw new ToolInputError("missing_option_value", "--input requires a path");
	if (args.output === "") throw new ToolInputError("missing_option_value", "--output requires a path");
	return args;
}

export function readToolInput(args) {
	const raw = args.input ? readFileSync(args.input, "utf8") : readFileSync(0, "utf8");
	if (!raw.trim()) return {};
	try {
		return JSON.parse(raw);
	} catch (error) {
		throw new ToolInputError("invalid_json", error instanceof Error ? error.message : String(error));
	}
}

export function okResult({ artifacts = [], warnings = [], data = null } = {}) {
	return { status: "ok", artifacts, warnings, errors: [], data };
}

export function errorResult(error) {
	const code = typeof error?.code === "string" ? error.code : "tool_error";
	const message = error instanceof Error ? error.message : String(error);
	return { status: "error", artifacts: [], warnings: [], errors: [{ code, message }] };
}

export function writeToolResult(result, args) {
	const text = `${JSON.stringify(result, null, 2)}\n`;
	if (args.output) writeFileSync(args.output, text);
	else process.stdout.write(text);
}

export async function runTool(handler) {
	let args;
	try {
		args = parseToolArguments();
		const input = readToolInput(args);
		writeToolResult(await handler(input), args);
	} catch (error) {
		writeToolResult(errorResult(error), args ?? { output: null });
		process.exitCode = 1;
	}
}

export function requiredString(input, key) {
	const value = input[key];
	if (typeof value !== "string" || value.trim() === "") {
		throw new ToolInputError("missing_required_field", `${key} is required`);
	}
	return value;
}

export function optionalInteger(input, key, fallback) {
	const value = input[key];
	if (value === undefined || value === null) return fallback;
	if (!Number.isInteger(value) || value < 0) {
		throw new ToolInputError("invalid_integer_field", `${key} must be a non-negative integer`);
	}
	return value;
}
